### ISSUE TITLE
The shipping information was unable to be saved. verify the input data and try again.

### ISSUE THREAD

**Issue by mehmetuygun** [magento/magento2#23908]
### Additional Details
Detailed explanation and result of the investigation  https://github.com/magento/magento2/issues/23908#issuecomment-538087658

### Preconditions (*)

1. Magento 2.3.2, 2.3-develop

### Steps to reproduce (*)

1. Create a customer account
2. Go to my account, and add a shipping address
2.5. Enable Persistent Shopping Cart **(updated Oct 10.2019)**
3. Add any product and ~go to the checkout~ **_!Impotent:_** wait until the customer session expires (one hour by default) but no longer than Persistence Lifetime is set to **(updated Oct 10.2019)**
3.5 Reload the page so that somewhere in the upper-right corner You get "Not You?" link next to the welcome message  **(updated Oct 10.2019)**
3.6 Clik on "Not You?" enter the credentials of the customer and log in **(updated Oct 10.2019)**
3.6 Go to Mini Cart end press Precede to Checkout **(updated Oct 10.2019)**
4. Select the address you have just added
5. Select a shipping method and click next button

### Expected result (*)

1. Payment page should show up

### Actual result (*)

1.  
![Screenshot_2](https://user-images.githubusercontent.com/4094099/61961203-f163b000-afbe-11e9-98b0-121b27b8d7b8.png)

**Comment by mehmetuygun** [comment]
@magento give me 2.3.2

**Comment by mehmetuygun** [comment]
@magento give me 2.3.2 instance

**Comment by magento-engcom-team** [comment]
Hi @mehmetuygun. Thank you for your request. I'm working on Magento 2.3.2 instance for you

**Comment by magento-engcom-team** [comment]
Hi @mehmetuygun, here is your [Magento instance](https://i-23908-2-3-2.instances.magento-community.engineering).
Admin access: [https://i-23908-2-3-2.instances.magento-community.engineering/admin](https://i-23908-2-3-2.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by engcom-Bravo** [comment]
Hello @mehmetuygun. Thanks for applying. We've been not able to reproduce Your issue on a vanilla magento instance. Please confirm You are able to to so.

**Comment by mehmetuygun** [comment]
I couldn't reproduce issue on magento instance either.
After I adding `$customerId = $this->customerSession->getId();` in to begging of the function `private function doValidate(AddressInterface $address, ?int $customerId): void` in the file `magento/module-quote/Model/QuoteAddressValidator.php` the error stop showing up. And I could pass to payment page. How is possible customerId passing as null? it happens when I logged in and trying place order with saved address.

**Comment by engcom-Bravo** [comment]
@mehmetuygun, if an issue is not reproduced on a fresh magento instance, free from any custom code, third party extensions, etc., then it is not a core magento bug this repository is intended for reporting. Please refer to the [Community Forums](https://community.magento.com/) or the [Magento Stack Exchange](https://magento.stackexchange.com/) site for advice or general discussion about this issue. Thus we are supposed to close this issue but please feel free to reopen it if deem necessary. Thanks for applying.

**Comment by ron-jomashop** [comment]
Looks like this ticket: https://github.com/magento/magento2/issues/8544

**Comment by ron-jomashop** [comment]
Could be related to this ticket: https://github.com/magento/magento2/issues/22112

**Comment by ron-jomashop** [comment]
This issue is related to this issue: https://github.com/magento/magento2/issues/23618

**Comment by onepack** [comment]
I created https://github.com/magento/magento2/issues/23618# and there is a reference to this issue. So I check and the first I see is that this issue is closed here as well.
I understand that you are eager to close issues to keep numbers low but clearly this issue can have a huge impact: Customers can not checkout.
Is there anyone from the magento core team that is willing to take time to find out what is the root of this issue as it is not entirely clear why people are getting this message.

**Comment by sbrendtro** [comment]
We've been seeing this as well on a 2.3.2 install.  Our data was imported from a Magento 1.9.x install using the Magento Data Migration Tool.  We were originally suspecting something to do with the data migration, but it still happens when you create a new account.  At the moment we're disabling our 3rd party modules one by one to see if we can narrow it down, but this behavior seems to have been around since before we added our 3rd party modules.

**Comment by onepack** [comment]
I imported the customer data via CSV as I have bad experiences with the data migration tool.
The account that had problems was a testing account. So I deleted it and re-created it. But it happened again after that with the new testing account (with the same email).

**Comment by sbrendtro** [comment]
Disabled all third party modules and still seeing same behavior.  Traced things back through the code and found some interesting things happening:

exception.log shows the error coming from 
[2019-08-06 22:05:07] main.CRITICAL: Invalid customer address id 160022 {"exception":"[object] (Magento\\Framework\\Exception\\NoSuchEntityException(code: 0): Invalid customer address id 160022 at .... vendor/magento/module-quote/Model/QuoteAddressValidator.php:79)"}

That's `QuoteAddressValidator::doValidate(AddressInterface $address, ?int $customerId)`, which is being called by line 130 in the same file.  The second parameter for doValidate is set as: `$cart->getCustomerIsGuest() ? null : $cart->getCustomer()->getId()`

However, $cart->getCustomerIsGuest() is reporting me as a guest, when in fact I'm not.  I logged in during checkout, and I'm continuing to refresh the checkout page, so it should be pretty apparent by now that I'm logged in.  It is actually showing me all of my addresses, so it clearly thinks I'm logged in.

Ultimately, getCustomerIsGuest() pulls the information from the quote table, which I checked and found that for my customer_id, my active quote had customer_is_guest column set to 1. I'm not sure how the quote (merged from my guest cart) ended up with customer_is_guest still set to 1, but after clearing it, the problem vanished (at least for now).

The following query helped me find the other quotes in the same situation:
```
SELECT entity_id, customer_id FROM quote WHERE customer_id != 0 AND customer_is_guest = 1;
```
Then I reset them to the proper status:
```
UPDATE quote SET customer_is_guest = 0 WHERE customer_id != 0 AND customer_is_guest = 1;
```

The only quotes I could find in this state belonged to test accounts.  I'll continue with testing with 3rd party modules enabled, and comment back if I find anything else improperly messing with the customer_is_guest field.

**Comment by onepack** [comment]
@sbrendtro, this is a good catch! I was not getting this critical message. But I was having the same in the quote table customer_is_guest was 1 with this testing customer. Setting it to 0 fixed things.
The root cause is yet unclear. I will keep an eye on this.

**Comment by ghost** [comment]
@sbrendtro 

Do you mean 

`UPDATE quote ...
`

rather than 

`UPDATE entity ...`

?

**Comment by sbrendtro** [comment]
> @sbrendtro
> 
> Do you mean
> 
> `UPDATE quote ... `
> 
> rather than
> 
> `UPDATE entity ...`
> 
> ?

Yes, thanks for pointing that out.  Just edited my comment.

**Comment by ghost** [comment]
No, thank you for pointing out the fix for what was turning out to be a massive pain in the backside for a client!

**Comment by ron-jomashop** [comment]
> @sbrendtro, this is a good catch! I was not getting this critical message. But I was having the same in the quote table customer_is_guest was 1 with this testing customer. Setting it to 0 fixed things.
> The root cause is yet unclear. I will keep an eye on this.

I think you should put this solution in your issue too because I believe it will help many people including me.

**Comment by ron-jomashop** [comment]
@onepack I think you should put this solution in your issue too because I believe it will help many people including me.

**Comment by mertgokceimam** [comment]
This is an issue that effects both Magento 2.2.9 and 2.3.2

It is also pretty disturbing to see Magento team denies that such issue exist. I can also confirm that the issue occurs on a clean installation. 

Thanks to @sbrendtro we corrected the issue on our shop but denying such important issues on an Enterprise solution is not acceptable on magento teams behalf.

**Comment by engcom-Bravo** [comment]
Dear members of this conversation. Please provide the steps following which we can reproduce the issue on a vanilla Magento installation of 2.3-develop version - the issue will be immediately reopened and confirmed. Dear @onepack, we are not eager to close the issues. We just follow our workflow (please see details to step 4 of this comment above https://github.com/magento/magento2/issues/23908#issuecomment-515489471). Thank You all for understanding.

**Comment by mertgokceimam** [comment]
It is really hard to reproduce as this is not an existing issue for everyone. For example when i checked our production instance , we had 85 total user quote that was effected by the issue. We of course corrected the quote issue but a step by step way to reproduce it is extremely hard. 

One day after correcting the quote's we have this issue back in place. 
`mysql> SELECT entity_id, customer_id FROM quote WHERE customer_id != 0 AND customer_is_guest = 1;
+-----------+-------------+
| entity_id | customer_id |
+-----------+-------------+
|     57600 |      313840 |
|     24340 |      397145 |
|    106394 |      483487 |
|    117654 |      483683 |
+-----------+-------------+`

Magento seriously urgently release a patch for this issue as this is effecting many customers incomes and most are not possibly aware of the issue.

Just because Magento team choose to close close issues for keeping clean board doesn't mean that us Enterprise customers are going to loose money.

In no condition Magento should allow a Logged in customer quote assigned as a Guest customer. But this can happen because of . a weird logic Magento team choose to implement. This shouldn't happen at all.

**Comment by tuyennn** [comment]
@engcom-Bravo We have same problem with magento since the 1st day of launching. The magento team should take a look on how you implement quote validation with `customer_is_guest` on quote and care how it's merging.

**Comment by sbrendtro** [comment]
Here's a temporary workaround at the database level (USE WITH CAUTION).  Since no one seems to be able to figure out where this is coming from, I figured I would use a MySQL trigger to prevent `customer_is_guest` from being set anything but 0 when `customer_id` is set.  You'll want to do your own testing before putting this in production.  But I can't think of any real risk, as `customer_is_guest` should really never be set to true for a logged-in customer.... right?

```
DROP TRIGGER IF EXISTS quote_before_insert_fix_guest;
DROP TRIGGER IF EXISTS quote_before_update_fix_guest;

DELIMITER $$

    CREATE TRIGGER quote_before_insert_fix_guest BEFORE INSERT ON quote
    FOR EACH ROW BEGIN
      IF (NEW.customer_id != 0 AND NEW.customer_is_guest != 0 ) THEN
            SET NEW.customer_is_guest = 0;
      END IF;
    END$$

    CREATE TRIGGER quote_before_update_fix_guest BEFORE UPDATE ON quote
    FOR EACH ROW BEGIN
      IF (NEW.customer_id != 0 AND NEW.customer_is_guest != 0 ) THEN
            SET NEW.customer_is_guest = 0;
      END IF;
    END$$

DELIMITER ;
```
edited: Both triggers were on BEFORE UPDATE, changed the first one to BEFORE INSERT.

**Comment by mertgokceimam** [comment]
@sbrendtro i will like to ask a question , do you also use Redis for session management ?

**Comment by tuyennn** [comment]
@mertgokceimam has same problem while using redis caching the session. Well, any investigation?

**Comment by sbrendtro** [comment]
No redis, we are using DB session storage.

**Comment by onepack** [comment]
Same here, no redis session caching.

**Comment by angelomaragna** [comment]
I've been able ro replicate a weird issue, which may be realted to this one, but don't know the correct frontend steps to get this to happen.

You need:
- persistent cart enabled and an expired cart for your account
- a current quote as guest, with customer_is_guest set a 1

once you get to the checkout and you log in, your cart is merged.

At that point, when you try to remove an item from the cart, the QuoteAddressValidator will always fail.


I've replaced the final function validateForCart with a temporary, and it does allow now to remove items from the cart.
```
 public function validateForCart(CartInterface $cart, AddressInterface $address): void
    {
        $customer = $cart->getCustomer();
        $customerId = $customer ? $customer->getId() : null;
        $this->doValidate(
                $address,
                        $customerId
                );
    }
```

**Comment by onepack** [comment]
This is a great find and fits in the way I was working. As I was testing I would always have products in the cart and would be busy coding or do other stuff on the backend. This would result in a session expiration so I would be logged out. Logging in again would create this issue with my test user.
But this could also happen in real-life situations when customers get logged out and will log in again in order to checkout.

**Comment by santhoshnsscoe** [comment]
Steps to reproduce:
1. Add products to cart
2. Go to checkout
3. Add email id and login from the checkout page
4. After login, click the next to go to the payment page.
5. Error will popup with "The shipping information was unable to be saved. Verify the input data and try again."

When we check the quote table.
```
SELECT entity_id, customer_id FROM quote WHERE customer_id != 0 AND customer_is_guest = 1;
+-----------+-------------+
| entity_id | customer_id |
+-----------+-------------+
|        28 |           4 |
+-----------+-------------+
```

**Comment by tcmax76** [comment]
We're experiencing this on our sites, as well. The database solution @sbrendtro fixed our problem accounts for now, but we'd really like a more permanent solution.

**Comment by engcom-Bravo** [comment]
@santhoshnsscoe, do I miss something:
![order](https://user-images.githubusercontent.com/51680745/64527114-bff43900-d30d-11e9-9454-da61e46761b9.gif)
? Thank You.

**Comment by tcmax76** [comment]
We had three more records show up in the quote table since Saturday marked with 1 for "customer_is_guest" but also had a customer_id entered (which seems to cause the error). I don't know if this makes a difference, but all three of them had more than an hour between the quote's 'created_at' datetime and the quote's 'updated_at' datetime.

**Comment by tcmax76** [comment]
Actually, there seems to be a correlation between these quotes that are erroring and having the 'is_persistent' column checked. We don't have a whole lot of quotes marked 'is_persistent', but all of the ones that include a customer_id while also marked 1 for 'customer_is_guest' (ie, the ones that seem to be causing this error) are marked as 'is_persistent'.

**Comment by santhoshnsscoe** [comment]
> @santhoshnsscoe, do I miss something:
> ![order](https://user-images.githubusercontent.com/51680745/64527114-bff43900-d30d-11e9-9454-da61e46761b9.gif)
> ? Thank You.

I checked with base Magento installation, it is working fine. So, there is some problem in either external modules or some settings. Will test completely and will update here.

**Comment by mertgokceimam** [comment]
Make sure to enable persistant cart so you can re produce it.

**Comment by webusat** [comment]
@sbrendtro I had the same issue and changing customer_is_guest to a 0 solved it for me.  Seems to be very sporadic and happened to two users.  It also seems related to the persistent cart settings.

**Comment by hostep** [comment]
We had to debug this issue as well today on a 2.3.1 shop, thanks to all the helpful comment above and our investigation, this is a brief summary:

- The issue here is that a logged in customer still can have the flag `customer_is_guest` set to 1 on the `quote`
- This flag is the root cause of the bug reported here, although it's not clear yet how this flag is being set incorrectly
- We've searched through various shops we maintain ranging between Magento 2.1, 2.2 and 2.3 versions and all versions are affected by this inconsistent `customer_is_guest` flag on the quotes (we've seen shops where hundreds of quotes existed where this is incorrect)
- We believe the bug reported here only came to the surface in Magento 2.3.1 an 2.3.2 due to some [changes done recently](https://github.com/magento/magento2/commit/73b6a0a2e683896e4f1b4a90b8929cf576f4c7f9#diff-4a45c227bcc0c1df3ba27e6563a656da) in the QuoteAddressValidator class. Before those versions, the incorrect flag also existed but didn't really seem to have caused issues.

Our current solution, was to add an after plugin to the `getCustomerIsGuest` function of the `Magento\Quote\Model\Quote` class. And do something like this:
```php
    public function afterGetCustomerIsGuest(QuoteModel $subject, $result)
    {
        $customer = $subject->getCustomer();

        return $customer->getId() === null;
    }
```

So instead of checking the flag from the database, we check if the customer's id is null or not to decide if a customer is a guest or not.

Now, this is only a workaround.
To fix this properly in Magento, two things need to happen in my opinion:
- the root cause of this bug needs to be found and fixed (https://github.com/magento/magento2/issues/24736 might be related)
- a database upgrade script should be written which fixes all incorrect `customer_is_guest` flags in the `quote` table, so all existing quotes which are being prevented to go through the checkout start working again.

I'm reopening this issue for now again (yes, I realise we still don't have steps to reproduce, but the issue seems to be very real).

**Comment by engcom-Bravo** [comment]
Hi everyone. Incredible but I've just got reproduced the issue on a (n almost) vanilla. Need some additional recheck - not yet sure for 100% I'll have it reproduced once more. But anyway I'll be back with the results.

**Comment by engcom-Bravo** [comment]
Updated Steps to reproduce. In fact, as I see now these steps to reproduce are scattered throughout the conversation. So theoretically I might have already gotten them if I were a bit more attentive (I bring my apologies for not being that) and it wouldn't have taken me that much time to come to it by myself. It seems to me I managed to reproduce the issue with slightly other steps as well but not regularly. The current steps will got the issue reproduced in any case.

**Comment by magento-engcom-team** [comment]
@engcom-Bravo Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-21568` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by hostep** [comment]
Great @engcom-Bravo !

Although, we've seen this problem happening on shops where the persistent cart functionality was never enabled, so there might be other ways to trigger this bug as well.
But it's good to already have one case documented properly 👍

**Comment by vijay-prevaj** [comment]
Is Anyone reproduce the issue in Magento-2.3.2 ?

Please List the steps to reproduce it.

**Comment by engcom-Bravo** [comment]
Hi, @vijay-prevaj. Have You already tried out [these ones](https://github.com/magento/magento2/issues/23908#issue-473403885)?

**Comment by diode-dynamics** [comment]
I'm not sure if this is appropriate, but can anyone provide further details of any workarounds that have been implemented right now? We are experiencing this issue on a daily basis, and it is tedious to run the script regularly as provided above. 

I see that someone added a plugin to a class, but I'm not sure how to implement that. If anyone has specific code I could add in anywhere until this is resolved, I would sincerely appreciate it. Thank you.

**Comment by hostep** [comment]
@diode-dynamics: A colleague of mine wrote the [following patch](https://github.com/magento/magento2/files/3746489/ISSUE-23908-Quote.txt), which you can apply using https://github.com/cweagans/composer-patches/ to the `magento/module-quote` composer package, this seems to work in our case.

Be aware that this is only a workaround, not a fix for the problem, the flag `customer_is_guest` in the database will still be incorrect from time to time, the patch simply causes that flag to be not used and determines the fact if a customer is a guest or not based on the customer id.

**Comment by diode-dynamics** [comment]
@hostep thank you so much! Understood that this is only a workaround. I've applied the patch, so far it seems to be working. This is a huge help for us until this bug is fixed. Thanks again.

**Comment by BenItt** [comment]
@hostep we are facing the same strange problem here. Unfortunately your Patch Link is leading to an empty page. Could you possibly repost your patch? That would be great. Thanks so much in advance

**Comment by vijay-prevaj** [comment]
> Hi, @vijay-prevaj. Have You already tried out [these ones](https://github.com/magento/magento2/issues/23908#issue-473403885)?

Hi @engcom-Bravo I tried it, but doesn't reproduce the issue in our magento instance.

Thank you

**Comment by sbrendtro** [comment]
@diode-dynamics We didn't want things to break if our patch got overwritten, so we created two database triggers to catch the improper `customer_is_guest` setting. I thought I posted this here, but apparently not. Remember, TEST, TEST, TEST before you add this to a production DB. Fixing in code is best, but until this fix is in a production release, I wanted to have a backup plan.

```
DELIMITER //

CREATE TRIGGER quote_before_insert_fix_guest BEFORE UPDATE ON quote
    FOR EACH ROW BEGIN
      IF (NEW.customer_id != 0 AND NEW.customer_is_guest != 0 ) THEN
            SET NEW.customer_is_guest = 0;
      END IF;
    END//

CREATE TRIGGER quote_before_update_fix_guest BEFORE UPDATE ON quote
    FOR EACH ROW BEGIN
      IF (NEW.customer_id != 0 AND NEW.customer_is_guest != 0 ) THEN
            SET NEW.customer_is_guest = 0;
      END IF;
    END//

DELIMITER ;
```

After adding these triggers, you will still need to reset the existing incorrect `customer_is_guest` values.

```
UPDATE quote SET customer_is_guest = 0 WHERE customer_id != 0 AND customer_is_guest = 1;
```

**Comment by BenItt** [comment]
@sbrendtro thank you for this additional information. We are facing the same problem and hope that Magento can fix this problem the soonest as it is really a very annoying bug.

**Comment by sdzhepa** [comment]
Hello @mehmetuygun  and @ffalkone 

I just got a notification that the current issue has been fixed by the internal Magento team and in the scope of `MC-21568`.
At this moment `MC-21568` has passed code-review and now in the QA stage. I expect it will be delivered during the first part of the next week.

FYI: @sbrendtro @onepack @gilesbennett @ron-jomashop @hostep @BenItt

**Comment by hostep** [comment]
Thanks @sdzhepa !

Is something like this being considered as well: https://github.com/magento/magento2/pull/24917#issuecomment-539666218 (which was based on @sbrendtro's comments above)?

Because fixing the bug in itself is not enough, incorrect data on the current quotes in the database needs to get fixed as well I think?

----

@BenItt: the patch download still works over here, but if it doesn't in your case for some reason, here is its content:
```diff
diff --git a/Model/Quote.php b/Model/Quote.php
index 6ac6e7715ad..6703a0cac42 100644
--- a/Model/Quote.php
+++ b/Model/Quote.php
@@ -691,6 +691,11 @@ class Quote extends AbstractExtensibleModel implements \Magento\Quote\Api\Data\C
      */
     public function getCustomerIsGuest()
     {
+        $customer = $this->getCustomer();
+        if ($customer) {
+            return is_null($customer->getId());
+        }
+
         return $this->_getData(self::KEY_CUSTOMER_IS_GUEST);
     }
 ```

**Comment by n2diving-net** [comment]
This appears to be a related cause https://github.com/magento/magento2/pull/25307

**Comment by hostep** [comment]
@n2diving-dgx: hmm, I don't agree, those are two separate issues in my opinion.

**Comment by magento-engcom-team** [comment]
Hi @mehmetuygun, @sdzhepa, @ffalkone. 

Thank you for your report and collaboration!

The issue was fixed by Magento team. The fix was delivered into `magento/magento2:2.3-develop` branch(es).
Related commit(s):
 - [6f0623f3f0f95ecdb960f5de89371fb8c8d51abc](https://github.com/magento/magento2/commit/6f0623f3f0f95ecdb960f5de89371fb8c8d51abc)

The fix will be available with the upcoming `2.3.4` release.

**Comment by sdzhepa** [comment]
Hello All
FYI: @mehmetuygun @ffalkone  @sbrendtro @onepack @gilesbennett @ron-jomashop @hostep @BenItt 

The internal Magento team has delivered the fix for this issue in the scope of `MC-21568`.
 Previously I asked them took into account for fix and testing this  https://github.com/magento/magento2/issues/23908#issuecomment-546057109 .

Now to be sure I would like to ask you for confirmation, is this fix OK for you? 
or some cases were missed and we need to create a new issue or re-open this one with additional information

Thank you in advance

**Comment by marmikdesai** [comment]
Hi,

I am facing this issue. Is this issue fixed?

**Comment by timelam** [comment]
> 
> 
> Hi @mehmetuygun, @sdzhepa, @ffalkone.
> 
> Thank you for your report and collaboration!
> 
> The issue was fixed by Magento team. The fix was delivered into `magento/magento2:2.3-develop` branch(es).
> Related commit(s):
> 
>     * [6f0623f3f0f95ecdb960f5de89371fb8c8d51abc](https://github.com/magento/magento2/commit/6f0623f3f0f95ecdb960f5de89371fb8c8d51abc)
> 
> 
> The fix will be available with the upcoming `2.3.4` release.


I had the same problem, but it was random and not present for every customer.

I removed line 92 from ./vendor/magento/module-persistent/Model/QuoteManager.php and the problem appears to be resolved.

https://github.com/magento/magento2/commit/6f0623f3f0f95ecdb960f5de89371fb8c8d51abc

**Comment by jaimiedijstra** [comment]
Magento 2.3.3 I have just come across this problem. Is there a step by step process to apply the fix wihle Magento are preparing 2.3.4 for launch mentioned in November? I'm using community edition. 

As others, this problem is a nightmare to reproduce, however by all description is the same and the main common issues:

- It was a testing account that I had a problem with checking out
- I ran `SELECT entity_id, customer_id FROM quote WHERE customer_id != 0 AND customer_is_guest = 1;` on the database and found it was only my own account cart, no other lines thank fully

A clear step by step process to temporarily fix this for a semi-technical magento user would be great!

**Comment by papanada** [comment]
Hi everybody,
(sorry for my poor english)
I doesn't understand if there is actually a solution for this issue...
There is someone (maybe @sbrendtro) that could help me ?

Thanks in advance !!!

**Comment by skitter-anton** [comment]
In my case it had to do with the php version. PHP 7.2 caused this issue. When we switched to 7.1 the problem did not occur anymore.

**Comment by mehmetuygun** [comment]
Hi @sdzhepa 

I would like to let you know @sdzhepa https://github.com/magento/magento2/issues/23908#issuecomment-546057109 This works for me. But After fixing database according to https://github.com/magento/magento2/pull/24917#issuecomment-539666218 (if the customer_id is not null in the quote table, customer_is_guest should be 0) But I don't need to make changes in codes.

With this https://github.com/magento/magento2/issues/23908#issuecomment-552400390 fix, I couldn't reproduce the issue again. 

I have applied this fix https://github.com/magento/magento2/issues/23908#issuecomment-552400390

Thank you.

**Comment by hunterallen** [comment]
> Hi @mehmetuygun, @sdzhepa, @ffalkone.
> 
> Thank you for your report and collaboration!
> 
> The issue was fixed by Magento team. The fix was delivered into `magento/magento2:2.3-develop` branch(es).
> Related commit(s):
> 
> * [6f0623f3f0f95ecdb960f5de89371fb8c8d51abc](https://github.com/magento/magento2/commit/6f0623f3f0f95ecdb960f5de89371fb8c8d51abc)
> 
> The fix will be available with the upcoming `2.3.4` release.

I am able to reproduce this in 2.3.4

**Comment by shopperwear** [comment]
I use magento 2.3.3 and the underlying problem seems to be with the url  persistent/index/unsetCookie/ when you use persistent shopping cart with remember me option enabled. When someone clicks not you? button at the top of a page, checkout page throws this error when saving address even if you log out and login again, it seems to persist.
A quick workaround, disable remember me option, clean storage cache and flush magento cache.

**Comment by snez** [comment]
@sdzhepa this issue is still present in Magento 2.3.4, what details can I provide to you for more investigation?

**Comment by snez** [comment]
Still a problem on Magento 2.3.4

![image](https://user-images.githubusercontent.com/249977/77757561-1884ec80-703a-11ea-81a0-268132910fe7.png)

Here is the full exception stack trace:

```
==> var/log/exception.log <==
[2020-03-27 12:44:06] main.CRITICAL: Invalid customer address id 2 {"exception":"[object] (Magento\\Framework\\Exception\\NoSuchEntityException(code: 0): Invalid customer address id 2 at /Users/snez/Projects/magento/m2-stripe-official.loc/vendor/magento/module-quote/Model/QuoteAddressValidator.php:79)"} []
[2020-03-27 12:45:38] main.INFO: #0 vendor/magento/module-quote/Model/QuoteAddressValidator.php(132): Magento\Quote\Model\QuoteAddressValidator->doValidate(Object(Magento\Quote\Model\Quote\Address\Interceptor), NULL)
#1 vendor/magento/module-checkout/Model/ShippingInformationManagement.php(172): Magento\Quote\Model\QuoteAddressValidator->validateForCart(Object(Magento\Quote\Model\Quote\Interceptor), Object(Magento\Quote\Model\Quote\Address\Interceptor))
#2 vendor/magento/framework/Interception/Interceptor.php(58): Magento\Checkout\Model\ShippingInformationManagement->saveAddressInformation(720, Object(Magento\Checkout\Model\ShippingInformation))
#3 vendor/magento/framework/Interception/Interceptor.php(138): Magento\Checkout\Model\ShippingInformationManagement\Interceptor->___callParent('saveAddressInfo...', Array)
#4 app/code/Swissup/Taxvat/Plugin/Model/ShippingInformationManagement.php(37): Magento\Checkout\Model\ShippingInformationManagement\Interceptor->Magento\Framework\Interception\{closure}(720, Object(Magento\Checkout\Model\ShippingInformation))
#5 vendor/magento/framework/Interception/Interceptor.php(135): Swissup\Taxvat\Plugin\Model\ShippingInformationManagement->aroundSaveAddressInformation(Object(Magento\Checkout\Model\ShippingInformationManagement\Interceptor), Object(Closure), 720, Object(Magento\Checkout\Model\ShippingInformation))
#6 vendor/magento/framework/Interception/Interceptor.php(153): Magento\Checkout\Model\ShippingInformationManagement\Interceptor->Magento\Framework\Interception\{closure}(720, Object(Magento\Checkout\Model\ShippingInformation))
#7 generated/code/Magento/Checkout/Model/ShippingInformationManagement/Interceptor.php(26): Magento\Checkout\Model\ShippingInformationManagement\Interceptor->___callPlugins('saveAddressInfo...', Array, NULL)
#8 [internal function]: Magento\Checkout\Model\ShippingInformationManagement\Interceptor->saveAddressInformation(720, Object(Magento\Checkout\Model\ShippingInformation))
#9 vendor/magento/module-webapi/Controller/Rest/SynchronousRequestProcessor.php(95): call_user_func_array(Array, Array)
#10 vendor/magento/module-webapi/Controller/Rest.php(188): Magento\Webapi\Controller\Rest\SynchronousRequestProcessor->process(Object(Magento\Framework\Webapi\Rest\Request\Proxy))
#11 vendor/magento/framework/Interception/Interceptor.php(58): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http))
#12 vendor/magento/framework/Interception/Interceptor.php(138): Magento\Webapi\Controller\Rest\Interceptor->___callParent('dispatch', Array)
#13 vendor/magento/framework/Interception/Interceptor.php(153): Magento\Webapi\Controller\Rest\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#14 generated/code/Magento/Webapi/Controller/Rest/Interceptor.php(26): Magento\Webapi\Controller\Rest\Interceptor->___callPlugins('dispatch', Array, Array)
#15 vendor/magento/framework/App/Http.php(116): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#16 generated/code/Magento/Framework/App/Http/Interceptor.php(24): Magento\Framework\App\Http->launch()
#17 vendor/magento/framework/App/Bootstrap.php(261): Magento\Framework\App\Http\Interceptor->launch()
#18 pub/index.php(40): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#19 {main} [] []
```

PHP version 7.2.27

**Comment by snez** [comment]
Still an issue on Magento 2.3.4

![image](https://user-images.githubusercontent.com/249977/77758043-fb9ce900-703a-11ea-9397-fc77286cb49d.png)

Here is the full stack trace:

```
[2020-03-27 12:54:45] main.CRITICAL: Invalid customer address id 2 {"exception":"[object] (Magento\\Framework\\Exception\\NoSuchEntityException(code: 0): Invalid customer address id 2 at /Users/snez/Projects/magento/m2-stripe-official.loc/vendor/magento/module-quote/Model/QuoteAddressValidator.php:81)"} []
[2020-03-27 12:55:13] main.INFO: #0 vendor/magento/module-quote/Model/QuoteAddressValidator.php(132): Magento\Quote\Model\QuoteAddressValidator->doValidate(Object(Magento\Quote\Model\Quote\Address), NULL)
#1 vendor/magento/module-checkout/Model/ShippingInformationManagement.php(172): Magento\Quote\Model\QuoteAddressValidator->validateForCart(Object(Magento\Quote\Model\Quote\Interceptor), Object(Magento\Quote\Model\Quote\Address))
#2 vendor/magento/framework/Interception/Interceptor.php(58): Magento\Checkout\Model\ShippingInformationManagement->saveAddressInformation(720, Object(Magento\Checkout\Model\ShippingInformation))
#3 vendor/magento/framework/Interception/Interceptor.php(138): Magento\Checkout\Model\ShippingInformationManagement\Interceptor->___callParent('saveAddressInfo...', Array)
#4 vendor/magento/framework/Interception/Interceptor.php(153): Magento\Checkout\Model\ShippingInformationManagement\Interceptor->Magento\Framework\Interception\{closure}(720, Object(Magento\Checkout\Model\ShippingInformation))
#5 generated/code/Magento/Checkout/Model/ShippingInformationManagement/Interceptor.php(26): Magento\Checkout\Model\ShippingInformationManagement\Interceptor->___callPlugins('saveAddressInfo...', Array, Array)
#6 [internal function]: Magento\Checkout\Model\ShippingInformationManagement\Interceptor->saveAddressInformation(720, Object(Magento\Checkout\Model\ShippingInformation))
#7 vendor/magento/module-webapi/Controller/Rest/SynchronousRequestProcessor.php(95): call_user_func_array(Array, Array)
#8 vendor/magento/module-webapi/Controller/Rest.php(188): Magento\Webapi\Controller\Rest\SynchronousRequestProcessor->process(Object(Magento\Framework\Webapi\Rest\Request\Proxy))
#9 vendor/magento/framework/Interception/Interceptor.php(58): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http))
#10 vendor/magento/framework/Interception/Interceptor.php(138): Magento\Webapi\Controller\Rest\Interceptor->___callParent('dispatch', Array)
#11 vendor/magento/framework/Interception/Interceptor.php(153): Magento\Webapi\Controller\Rest\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#12 generated/code/Magento/Webapi/Controller/Rest/Interceptor.php(26): Magento\Webapi\Controller\Rest\Interceptor->___callPlugins('dispatch', Array, Array)
#13 vendor/magento/framework/App/Http.php(116): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#14 generated/code/Magento/Framework/App/Http/Interceptor.php(24): Magento\Framework\App\Http->launch()
#15 vendor/magento/framework/App/Bootstrap.php(261): Magento\Framework\App\Http\Interceptor->launch()
#16 pub/index.php(40): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
```

PHP version 7.2.27

**Comment by gaokeji52** [comment]
I am running the site on version 2.3.4, I got the same error.

**Comment by vseager** [comment]
I can also verify that we are seeing this issue on 2.3.4

**Comment by hostep** [comment]
Thanks for the feedback @vseager!

@engcom-Bravo: are [the steps to reproduce](https://github.com/magento/magento2/issues/23908#issuecomment-539479231) still applicable on 2.3.4 or 2.4-develop? Or do we need to find other steps to reproduce the issue?

**Comment by Stoyvo** [comment]
Still seeing this on new instance of 2.3.4. No clear steps, but after a few attempts it does happen

**Comment by alfaproton** [comment]
Hi @hostep. Your patch prevents customer creation when you create an order from Admin. (Create New Order -> Create New Customer). 
I think it should be:
`$customer = $this->getCustomer();
  if ($customer && !$this->getIsSuperMode()) {
  return is_null($customer->getId());
  }`

**Comment by mahmoodkk** [comment]
> $customerId = $this->customerSession->getId();

This code worked for me.

**Comment by ihor-sviziev** [comment]
Re-opening this issue. seems like it was closed without any reason.
CC @sidolov

**Comment by tushhan** [comment]
The same problem persists after upgrading to Magento 2.4.1

**Comment by APanteleychuk** [comment]
I have same problem too. After upgrading to Magento 2.4.1

**Comment by oviliz** [comment]
Hopefully, this will be fixed within two years... :(

**Comment by snez** [comment]
For people still struggling with this, if you delete the customer addresses from the Magento admin, the customer will then be able to use the checkout. The issue is caused from data corruption when a guest customer logs in. Deleting the address fixes the corrupted data.

**Comment by ihor-sviziev** [comment]
@sidolov @gabrieldagama seems like priority for this issue was set incorrectly, it should be p2 or even p1 as the customer can't place an order in such case

**Comment by sidolov** [comment]
@ihor-sviziev changed to P1, thank you for the clarification!

**Comment by jpillai88** [comment]
Hi,
What is the status of this issue ?
Is there any version of magento resolve this issue?

**Comment by willryanuk** [comment]
This is still not resolved in Magento 2.4.2 CE.

**Comment by groomershop-mt** [comment]
We can also confirm that this problem still occur at Magento 2.4.2 CE - about 0,2% of records in our quote table has set customer_is_guest=1 for record having not null customer_id. This problems makes some clients unable to place an order, as it raise exception at:

/app/m/vendor/magento/module-quote/Model/QuoteAddressValidator.php in Magento\Quote\Model\QuoteAddressValidator::doValidate at line 79

        }
        if ($address->getCustomerAddressId()) {
            //Existing address cannot belong to a guest
            if (!$customerId) {
                **throw new \Magento\Framework\Exception\NoSuchEntityException(
Link your stack trace to your source code.
                    __('Invalid customer address id %1', $address->getCustomerAddressId())
                );**
            }
            //Validating address ID
            try {

**Comment by katonahmike** [comment]
We also have this issue after upgrading to Magento 2.4.2. Support is claiming it's related the 3rd party extensions. It is affecting all existing customers who make the mistake of logging-in.

**Comment by paugnu** [comment]
Hi all,

I've found an important issue that could be related to what is happening here:

This is the scenario:
1. A guest customer adds products to the cart and proceeds to checkout
2. Introduces address data, selects shipping and payment methods.
3. Registration is mandatory, so the customer completes also password fields
4. Now clicks on 'Complete the order' and is redirected to the payment gateway (this might depend on the payment method).
-- At this moment, a new order and customer are created
-- Order is in pending status
5. If the payment fails, the customer goes back to the cart.
-- Order is in cancelled status

To restore the customer cart, we use in our module the Magento native function:
vendor/magento/module-checkout/Model/Session.php::restoreQuote()

Which actually restores the customer quote based on the quote that it was being use for the order. And here is the problem:
- Quote is restored with original values. This means:
-- customer_id column is empty, even though when the customer is redirected back to the cart is already logged
-- customer_is_guest is set to 1, when this is not valid anymore
-- checkout_method is set to 'guest', when it should be now 'NULL'
-- customer_group_id is 0, when it should have the new customer group id

Because of these incorrect values, the customer can't complete a new order. It's stuck because quote data is inconsistent.

A possible fix for this particular case could be replacing the restoreQuote() function in vendor/magento/module-checkout/Model/Session.php, with the following. But note that it would be necessary to add Magento\Customer\Model\Session as one of the dependencies:

```
public function restoreQuote()
    {
        /** @var \Magento\Sales\Model\Order $order */
        $order = $this->getLastRealOrder();
        if ($order->getId()) {
            try {
                $quote = $this->quoteRepository->get($order->getQuoteId());
                $quote->setIsActive(1)->setReservedOrderId(null);

                if ($this->customerSession->isLoggedIn()) {
                    $quote->setData('customer_id', $this->customerSession->getCustomer()->getId());
                    $quote->setData('customer_is_guest', 0);
                    $quote->setData('checkout_method', 'NULL');
                    $quote->setData('customer_group_id', $this->customerSession->getCustomer()->getGroupId());
                }

                $this->quoteRepository->save($quote);
                $this->replaceQuote($quote)->unsLastRealOrderId();
                $this->_eventManager->dispatch('restore_quote', ['order' => $order, 'quote' => $quote]);
                return true;
            } catch (NoSuchEntityException $e) {
                $this->logger->critical($e);
            }
        }

        return false;
    }
```

**Comment by ralex78** [comment]
> We also have this issue after upgrading to Magento 2.4.2. Support is claiming it's related the 3rd party extensions. It is affecting all existing customers who make the mistake of logging-in.

I can confirm this issue on CE 2.4.2. I'm hardly using any 3rd party extensions.

**Comment by Ashraf-alkibsi** [comment]
for me setting the default currency and country solved this issue.

**Comment by ralex78** [comment]
> for me setting the default currency and country solved this issue.

Not working in my case. Default currency and country are both set.

**Comment by Ashraf-alkibsi** [comment]
@ralex78 try checking the log files, I hope it can give you a hint of what causing the problem

**Comment by ralex78** [comment]
> @ralex78 try checking the log files, I hope it can give you a hint of what causing the problem

Unfortunately, I am not a developer and not even quite tech-savvy, but I'll try this and see if it will give me any hints. Thanks.

**Comment by pawelskowronek** [comment]
I have come across the same issue and following @Ashraf-alkibsi advice I have checked exception.log. It turned out that I haven't set a Currency Rate for the new Currency.

You can set the currency rate by going to **Stores -> Currency -> Currency Rates**. 

Here is the error I was getting in the exception.log
`report.CRITICAL: Undefined rate from "USD-GBP".`

**Comment by waldroot** [comment]
If someone still struggles with this issue I fixed it by modifying the class \Magento\Quote\Model\Quote. Maybe it's not the best but it works.

`--- vendor/magento/module-quote/Model/Quote.php
+++ vendor/magento/module-quote/Model/Quote.php
@@ -853,8 +853,11 @@
             $this->setIsChanged(0);
         }
 
-        if ($this->_customer) {
+        if ($this->_customer && $this->_customer->getId()) {
             $this->setCustomerId($this->_customer->getId());
+            $this->setCustomerIsGuest(0);
+        }else{
+            $this->setCustomerIsGuest(1);
         }
 `
Still didn't figured out how to replicate it, we had all currency and rates correctly set and until the fix was applied we encountered constantly this problem.

**Comment by moranotcsos** [comment]
Hello,

Regarding https://github.com/magento/magento2/issues/23908#issuecomment-518875486 and the customer_is_guest setting being set to 1 even though the customer is logged in, I was able to reproduce this issue, but I am not sure if this is the same for everyone.

What appears to be happening for me is that if a customer is a guest, places the order, and the order fails to be placed (for example, a hosted payment page), the quote's customer_is_guest remains set as true. If the customer then creates a new account, the quote gets assigned to that customer as expected, but the customer_is_guest setting never switches to false for some reason. After the customer creates the account, if they set a default shipping address, go to checkout, and then hit Next, the doValidate function tries to run, but since their default shipping address is assigned to the customer and their cart is marked as a "guest cart", the doValidate funtion fails on line 76 - 82:

`
if ($address->getCustomerAddressId()) {
    //Existing address cannot belong to a guest
    if (!$customerId) {
        throw new \Magento\Framework\Exception\NoSuchEntityException(
            __('Invalid customer address id %1', $address->getCustomerAddressId())
        );
    }
`

**Comment by lbajsarowicz** [comment]
I'm currently testing the solution from https://github.com/magento/magento2/issues/23618

**Comment by engcom-November** [comment]
Verified the issue on Magento 2.4-develop branch and the issue is not reproducible. Also, related JIRA is closed and [PR](https://github.com/magento-commerce/magento2ce/pull/4998) is merged. Hence we are closing this issue.
Thank you.

**Comment by oviliz** [comment]
@engcom-November this is a popular cross-version issue going around for many years and you took the liberty to close it simply because you were unable to reproduce?!... :(

**Comment by ihor-sviziev** [comment]
@engcom-November, could you provide some **public** pull request link or links to commits that fix this issue?

**Comment by hostep** [comment]
@sidolov: can you help in providing more public info, for example a list of commits that fixed this, like Ihor requested? I'm also interested in this. Thanks!

**Comment by sidolov** [comment]
@hostep , @ihor-sviziev as far as I can see the issue was fixed in scope of commit: https://github.com/magento/magento2/commit/6f0623f3f0f95ecdb960f5de89371fb8c8d51abc

**Comment by ihor-sviziev** [comment]
@sidolov thank you!

**Comment by hostep** [comment]
@sidolov: thanks, but I'm pretty sure this bug isn't fixed. Because that commit you referenced was introduced in Magento 2.3.4, even though the bug still exists in newer Magento versions ...

**Comment by sidolov** [comment]
@hostep no problem, let's re-open it and check once again. Thanks!

**Comment by bartoszkubicki** [comment]
Still an issue on newest magento version

**Comment by engcom-November** [comment]
Verified the issue on Magento 2.4-develop branch but cannot able to reproduce the issue due to below issue:;
https://github.com/magento/magento2/issues/35182 - "Not You?" link is not getting displayed after customer session is expired to verify this issue.

**Comment by redati** [comment]
Magento 2.4.4


[2022-04-13T12:52:34.090875+00:00] main.CRITICAL: Magento\Framework\Exception\NoSuchEntityException: Id do endereço inválido 10781 in /var/www/html/vendor/magento/module-quote/Model/QuoteAddressValidator.php:79
Stack trace:
#0 /var/www/html/vendor/magento/module-quote/Model/QuoteAddressValidator.php(130): Magento\Quote\Model\QuoteAddressValidator->doValidate()
#1 /var/www/html/vendor/magento/module-checkout/Model/ShippingInformationManagement.php(179): Magento\Quote\Model\QuoteAddressValidator->validateForCart()
#2 [internal function]: Magento\Checkout\Model\ShippingInformationManagement->saveAddressInformation()
#3 /var/www/html/vendor/magento/module-webapi/Controller/Rest/SynchronousRequestProcessor.php(95): call_user_func_array()
#4 /var/www/html/vendor/magento/module-webapi/Controller/Rest.php(188): Magento\Webapi\Controller\Rest\SynchronousRequestProcessor->process()
#5 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Webapi\Controller\Rest->dispatch()
#6 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Webapi\Controller\Rest\Interceptor->___callParent()
#7 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Webapi\Controller\Rest\Interceptor->Magento\Framework\Interception\{closure}()
#8 /var/www/html/generated/code/Magento/Webapi/Controller/Rest/Interceptor.php(23): Magento\Webapi\Controller\Rest\Interceptor->___callPlugins()
#9 /var/www/html/vendor/magento/framework/App/Http.php(116): Magento\Webapi\Controller\Rest\Interceptor->dispatch()
#10 /var/www/html/vendor/magento/framework/App/Bootstrap.php(264): Magento\Framework\App\Http->launch()
#11 /var/www/html/pub/index.php(30): Magento\Framework\App\Bootstrap->run()
#12 {main} [] []

**Comment by engcom-Delta** [comment]
We have noticed that this issue has not been updated for a period of more than 14 Days. 
Hence we assume that this issue is fixed now, so we are closing it. Please raise a fresh ticket or reopen this ticket if you need more assistance on this.

**Comment by oviliz** [comment]
Incredible...

**Comment by niners52** [comment]
Still happens on 2.4.5

**Comment by HansMerz** [comment]
Still happens on 2.4.6-p2

**Comment by oviliz** [comment]
@engcom-Delta this is obviously unresolved... could you please reopen?

**Comment by onimitch** [comment]
Hello maybe this will help.

I just got this error while shopping on www.lindt.co.uk. I obviously can't tell you what version they are using but following the error lead me here. 

This is what I did:

1. Shopping for a product and added to my cart. I had no account yet, so just anon.
2. Went to cart and checkout.
3. Created a new account using login/register with Google.
4. I think had an issue where I couldn't checkout because it said there were no products even though they were listed (the no products error occured when I tried to enter a promotional code).
5. I cleared the cart and re-added my product.
6. Went to cart, added my promotional code and did checkout.
7. Any address I select in shipping information gives the error "The shipping information was unable to be saved. Error: Invalid customer address id XXXX" where XXXX is an ID.

In the end I had to give up with lindt and order as a guest, but maybe the above steps will help to reproduce the issue.
As a developer (who doesn't use magento I should add though), my suspicion is the problem might have occured when registering using google.

Good luck!

**Comment by vandark25** [comment]
Solved in https://github.com/magento/magento2/issues/23618
